java/java.hints/src/org/netbeans/modules/java/hints/perf/Tiny.java

```diff
@@ -462,36 +485,94 @@ public static ErrorDescription redundantToString(HintContext ctx) {
         suppressWarnings = "UnnecessaryTemporaryOnConversionFromString" 
     )
     public static ErrorDescription unnecessaryTempFromString(HintContext ctx) {
-        TypeMirror resType = ctx.getInfo().getTrees().getTypeMirror(ctx.getPath());
-        if (resType == null) {
-            return null;
+
+        String type;
+        String method;
+
+        // determine if the destination is primitive
+        TypeMirror destType = getDestinationType(ctx, ctx.getPath());
+        TypeMirror srcType = ctx.getInfo().getTrees().getTypeMirror(ctx.getPath());
+
+        if (destType instanceof PrimitiveType && !(srcType instanceof PrimitiveType)) {
+            srcType = ctx.getInfo().getTypes().unboxedType(srcType);
+            String[] replacement = PARSE_METHODS.get(srcType.getKind());
+            type = replacement[0];
+            method = replacement[1];
+        } else if (!(destType instanceof PrimitiveType) && srcType instanceof PrimitiveType) {
+            type = PARSE_METHODS.get(srcType.getKind())[0];
+            method = "valueOf";  // NOI18N
+        } else {
+            return null;  // nothing to do, a different rule handles .intValue() boxing problems
         }
-        if (resType.getKind() == TypeKind.BOOLEAN) {
-            if (ctx.getInfo().getSourceVersion().compareTo(SourceVersion.RELEASE_5) < 0) {
-                // might alter new Boolean($v) to Boolean.valueOf($v), but that's all we can do. JDK < 5 has no 
-                // primitive-valued pasre* method for booleans.
-                return null;
-            }
+
+        if (srcType.getKind() == TypeKind.BOOLEAN && ctx.getInfo().getSourceVersion().compareTo(SourceVersion.RELEASE_5) < 0) {
+            return null;  // JDK < 5 has no primitive-valued pasre* method for booleans.
         }
-        String[] arr = PARSE_METHODS.get(resType.getKind());
-        if (arr == null) {
-            return null; // just in case
+
+        Fix fix = JavaFixUtilities.rewriteFix(ctx, Bundle.FIX_UnnecessaryTempFromString1(type, method), ctx.getPath(), type + "." + method + "($v)"); // NOI18N
+        return ErrorDescriptionFactory.forTree(ctx, ctx.getPath(), Bundle.TEXT_UnnecessaryTempFromString(), fix); // NOI18N
+    }
+
+    private static TypeMirror getDestinationType(HintContext ctx, TreePath path) {
+
+        TreePath parent = path.getParentPath();
+        Tree parentLeaf = parent.getLeaf();
+
+        Types types = ctx.getInfo().getTypes();
+        Trees trees = ctx.getInfo().getTrees();
+
+        if (parentLeaf instanceof MethodInvocationTree) {
+
+            MethodInvocationTree met = (MethodInvocationTree) parentLeaf;
+
+            int index = met.getArguments().indexOf(path.getLeaf());
+
+            TypeMirror method = trees.getElement(new TreePath(path, met.getMethodSelect())).asType();
```

Comment:
As a general note - doing `Element.asType()` will return the type of the generic declaration (i.e. a type with the generic type parameters, like e.g. `(E)boolean` for `List.add`). What is likely needed here is the actual types used for the method invocation, with generic types (if any) resolved (e.g. `(Integer)boolean` for `List<Integer> l; l.add(...)`). `trees.getTypeMirror(<method-select>)` should return that, AFAIK.

(Note there's a slight inconsistency/difference between the `getElement` and `getTypeMirror` methods:
-if the method select is passed into the methods, they will return the `ExecutableElement` and `ExecutableType` for the invocation
-if the method invocation is passed into the methods, `getElement` will still return the `ExecutableElement`, but `getTypeMirror` will return the actual return type (with all generics resolved as needed)
)

Reply:
I tried getTypeMirror first actually and it worked for methods, but I could not figure out how to make the NewClassTree work too. So instead of using two ways of handling very similar branches of an if, i decided to solve both the same.

The current version I just committed uses getTypeMirror for the method branch and Element.asType for the class branch. I would need some help to make both use getTypeMirror (i am missing the getMethodSelect() equivalent for constructors) - since i could not figure out how.

Reply:
the new code:
https://github.com/apache/netbeans/blob/96f3355749c8500a53cae56a98abe2288887521a/java/java.hints/src/org/netbeans/modules/java/hints/perf/Tiny.java#L521-L531

Reply:
@jlahoda you think it is OK to leave the NEW_CLASS if-branch as is? I still don't know how to get a handle on a constructor, with getTypeMirror I would only get the class type, not the constructor type with its parameters.

Reply:
@jlahoda ping :)